Detectors/FIT/FV0/simulation/src/Digitizer.cxx

```diff
@@ -116,56 +122,65 @@ void Digitizer::process(const std::vector<o2::fv0::Hit>& hits,
 
     Double_t const nPhotons = hitEdep * DigitizationParameters::N_PHOTONS_PER_MEV;
     Int_t const nPhE = SimulateLightYield(detId, nPhotons);
-    Float_t const t = mScintillatorDelay.at(mIteratorScintDelay) + hit.GetTime() * 1e9;
-    Float_t charge = TMath::Qe() * DigitizationParameters::mPmtGain * mBinSize / mPmtTimeIntegral;
+    Float_t const t = mRndScintDelay.getNextValue() + hit.GetTime() * 1e9;
+    Float_t const charge = TMath::Qe() * DigitizationParameters::mPmtGain * mBinSize / mPmtTimeIntegral;
+
+    auto& analogSignal = mPmtChargeVsTime[detId];
+
     for (Int_t iPhE = 0; iPhE < nPhE; ++iPhE) {
-      Float_t const tPhE = t + mSignalShape.at(mIteratorSignalShape);
-      Int_t const firstBin = TMath::Max((Int_t)0, (Int_t)((tPhE - DigitizationParameters::mPmtTransitTime) / mBinSize));
+      Float_t const tPhE = t + mRndSignalShape.getNextValue();
+      Int_t const firstBin = roundVc(TMath::Max((Int_t)0, (Int_t)((tPhE - DigitizationParameters::mPmtTransitTime) / mBinSize)));
       Int_t const lastBin = TMath::Min((Int_t)mNBins - 1, (Int_t)((tPhE + 2. * DigitizationParameters::mPmtTransitTime) / mBinSize));
-      for (Int_t iBin = firstBin; iBin <= lastBin; ++iBin) {
-        Float_t const tempT = mBinSize * (0.5 + iBin) - tPhE;
-        mPmtChargeVsTime.at(detId).at(iBin) += mGainVar.at(mIteratorGainVar) * charge * PmtResponse(tempT);
-      } // time-bin loop
-
-      // RandomRing servicing (perhaps a separate class could do better)
-      mIteratorGainVar++;
-      if (mIteratorGainVar >= mGainVar.size()) {
-        mIteratorGainVar = 0;
+      Float_t tempT = mBinSize * (0.5f + firstBin) - tPhE;
+      Float_t* p = analogSignal.data() + firstBin;
+      Vc::float_v binVc(tempT);
+      binVc += mBinSize * Vc::float_v::IndexesFromZero();
+      Vc::float_v binIncVc(mBinSize * Vc::float_v::Size);
+
+      Vc::float_v workVc;
+      Int_t iBin = firstBin;
+      for (Int_t const lastBinRounded = roundVc(lastBin); iBin < lastBinRounded;) {
+        workVc.load(p);
+        workVc += mRndGainVar.getNextValueVc() * charge * PmtResponseVc(binVc);
+        workVc.store(p);
+        Vc::prefetchForOneRead(p);
+        binVc += binIncVc;
```

Comment:
So essentially it seems you are evaluating PmtResponseVc in discrete steps. Couldn't you hence just tabulate the response with this granularity? This should completely remove the exp function from the runtime.

Reply:
Hi @hcab14, since you were the one who implemented it, could you comment on this?

Reply:
One more idea (for the future): The parallelism is now over bins. Alternatively one could try making it parallel over photons which would increase the SIMD portion of the code.

Reply:
> So essentially it seems you are evaluating PmtResponseVc in discrete steps. Couldn't you hence just tabulate the response with this granularity? This should completely remove the exp function from the runtime.

I will have a look into this. For each photon there is a different offset smaller than the step size, so we need tables for a small number (_e.g._ 8) of offsets.

> One more idea (for the future): The parallelism is now over bins. Alternatively one could try making it parallel over photons which would increase the SIMD portion of the code.

There are currently 2000 bins and _O_(100) photons. The innermost loop runs only over a relevant subset of these bins between `binFirst` and `binLast`, so I am not sure if it will be faster to reverse the order of the loops because then one has to always loop over the 2000 bins.

Reply:
> So essentially it seems you are evaluating PmtResponseVc in discrete steps. Couldn't you hence just tabulate the response with this granularity? This should completely remove the exp function from the runtime.

I have implemented the table lookup and it is 3-4 times faster than the SIMD version.

Unfortunately I get segmentation faults during memory allocations, _e.g._, in std::vector::resize, and I am not sure where the problem is. Can it be due to a mismatch of dynamic libraries (I am using Ubuntu 19.10)? Also, for some reason I am not able to make o2 compile with clang.

Reply:
Cool. Is it part of this PR? 

(I could merge this one in any case and we can make a new one later. Do you have the code in some branch? I could try to take a look and investigate the segfault.)

Reply:
The look-up table is not part of this PR. I'd say it's better to merge this one once the tests are successful (there are other bug fixes here as well) and implement further optimization in the next PR.